Lib/test/test_ctypes/test_dlerror.py

```diff
@@ -0,0 +1,88 @@
+import os.path
+import sys
+import unittest
+from ctypes import CDLL
+
+FOO_C = r"""
+#include <stdio.h>
+
+/* This is a 'GNU indirect function' (IFUNC) that will be called by
+   dlsym() to resolve the symbol "foo" to an address. Typically, such
+   a function would return the address of an actual function, but it
+   can also just return NULL.  For some background on IFUNCs, see
+   https://willnewton.name/uncategorized/using-gnu-indirect-functions/
+
+   Adapted from Michael Kerrisk's answer: https://stackoverflow.com/a/53590014
+*/
+
+asm (".type foo, @gnu_indirect_function");
+
+void *foo(void)
+{
+    fprintf(stderr, "foo IFUNC called\n");
+    return NULL;
+}
+"""
+
+
+@unittest.skipUnless(sys.platform.startswith('linux'),
+                     'Test only valid for Linux')
+class TestNullDlsym(unittest.TestCase):
+    """GH-126554: Ensure that we catch NULL dlsym return values
+
+    In rare cases, such as when using GNU IFUNCs, dlsym(),
+    the C function that ctypes' CDLL uses to get the address
+    of symbols, can return NULL.
+
+    The objective way of telling if an error during symbol
+    lookup happened is to call glibc's dlerror() and check
+    for a non-NULL return value.
+
+    However, there can be cases where dlsym() returns NULL
+    and dlerror() is also NULL, meaning that glibc did not
+    encounter any error.
+
+    In the case of ctypes, we subjectively treat that as
+    an error, and throw a relevant exception.
+
+    This test case ensures that we correctly enforce
+    this 'dlsym returned NULL -> throw Error' rule.
+    """
+
+    def test_null_dlsym(self):
+        import subprocess
+        import tempfile
+
+        try:
+            p = subprocess.Popen(['gcc', '--version'], stdout=subprocess.PIPE,
+                                 stderr=subprocess.DEVNULL)
+            out, _ = p.communicate()
```

Comment:
AFACT, this variable is not used. Can we use someting else to check for the existence of `gcc`, e.g., `subprocess.run` + check the returncode?

Reply:
Would `subprocess.check_call` work?

Reply:
I always forget which one is the easiest to use but using:

```py
retcode = subprocess.call(["gcc", "--version"], 
					      stdout=subprocess.DEVNULL,
				          stderr=subprocess.DEVNULL,
					      shell=True)
if retcode != 0:
	self.skipTest("gcc is missing")
```

Reply:
Done and pushed.

Reply:
Actually, something might be off.

I get a `retcode` of 1, even though `gcc` is present, in the sense that if I remove the check, the lines below compile `libfoo.so` normally, and the test completes.

Reply:
You're right, here's what I got:

```pycon
>>> subprocess.call(["gcc", "--version"])
gcc (SUSE Linux) 7.5.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

0
>>> subprocess.call(["gcc", "--version"], shell=True)
gcc: fatal error: no input files
compilation terminated.
1
>>> subprocess.call(["gcc", "--version"], shell=True)
```

Reply:
Ok, so the `shell` parameter may not be necessary.

Reply:
It was the `shell=True` part. I logged the stderr, and I get `gcc: No input specified`.

This means that the `--version` argument was not passed correctly.

I'll investigate. We want to keep the `shell=True` part, but have it work as expected.

Reply:
Correct code:

```py
try:
	retcode = subprocess.call(["gcc", "--version"], 
	                          stdout=subprocess.DEVNULL,
	                          stderr=subprocess.DEVNULL)
except OSError:
	self.skipTest("gcc is missing")
else:
	if retcode != 0:
		self.skipTest("gcc --version failed"
```

Reply:
I don't necessarily think we need the `shell` part since we're anyway on Linux only. Btw, the reason why `--version` failed is because it expected `gcc --version` as a single string and not a list of arguments (though I don't why it seems to incorrectly dispatch the command).

Reply:
Figured it out.

When using `shell=True`, the command to run must be a single string, passed as the first argument.

Any additional arguments are passed to the shell itself.

So, in this case, we were running the equivalent of:

```
/bin/sh -c 'gcc' --version
```

Quoting the docs [1]:

> On POSIX with shell=True, the shell defaults to /bin/sh. If args is a string, the string specifies the command to execute through the shell. This means that the string must be formatted exactly as it would be when typed at the shell prompt. This includes, for example, quoting or backslash escaping filenames with spaces in them. If args is a sequence, the first item specifies the command string, and any additional items will be treated as additional arguments to the shell itself. That is to say, [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) does the equivalent of: `Popen(['/bin/sh', '-c', args[0], args[1], ...])`

[1] https://docs.python.org/3/library/subprocess.html#popen-constructor

Reply:
I'll merge into a single string, "gcc --version" and keep `shell=True`.

Reply:
@picnixz I'm on Firefox, and I wasn't seeing your comments in real time.

So the above, which kinda looks like a conversation, was a monologue on my side...

If we keep the `shell=True`, no need to catch the OSError.

If we remove the `shell=True`, we'll have to catch it.

Reply:
Let's keep `shell=True`, which is in line with the commit I pushed.

Reply:
>  I'm on Firefox, and I wasn't seeing your comments in real time for some reason.

We're both on FF then and I also thought it was a monologue.

Reply:
We reached the same conclusion any way :P